MelodyGenerator.js

```diff
@@ -0,0 +1,145 @@
+const durations = require('./durations.js');
+const MelodyQueue = require('./MelodyQueue.js');
+const waveGenerator = require('./WaveGenerator.js');
+const note = require('./note.js');
+
+//Global variables that are accessible to all
+let melody = new MelodyQueue();
+let macros = new Map(); // In case a macro needs to be defined
+const regex1 = /repeat (\d) times (.+)/; // Regex for repeat commands
+const regex2 = /((note\(|pause\()((.{2,3})(;(\w\w))*)(,\s?([A-Z]))?\))/; //Regex for notes, pauses
+const regex3 = /define (\w+) is (.+)/;  //Regex for defining macros
+const regex4 = /(play )?(.+)/; // Regex for playing non-repeating macros
+
+class MelodyGenerator
+{
+
+	parseMacro(lineString) {
+		// define mel01 is pause(HT) | note(E5, E)
+		const parsedMacro = regex3.exec(lineString); 
+		const macroName = parsedMacro[1];
+		const line = parsedMacro[2].split("|");
+		let macroNotes = [];
+
+		for(let i=0; i < line.length; i++) {
+				macroNotes = [...macroNotes, ...this.parseSound(line[i])];
+		}
+		macros.set(macroName, macroNotes);
+
+	}
+
+	parseSound(soundString) {
+		let soundTest = regex2.test(soundString); // Is it a note/pause? (It could be a repeat)
+		let sound = regex2.exec(soundString);
+		
+
+		// [2] contains 'pause' or 'note'
```

Comment:
I will have to disagree with you here because in that function I end up making references to sound[8], sound[3] as well as sound[2]. The key was for me not to have to keep looking it up.

So this is where I have to bring up a guideline that I use when coding and I'm not sure if I'm thinking about it the right way. My aim when coding is to use the least amount of code and for the computer to make any single calculation once. The idea is the same processing power and, although it's not at all an issue here, it's a good habit to have. Taking this into consideration, I'd rather have the computer calculate the value of `sound` once, and then refer to positions 2, 3, and 8 later rather than calculate it three times and just save the values as separate vars. Do you see what I mean? Do you agree with my premise or is this something I should rethink?